core/src/main/java/org/infinispan/interceptors/impl/BaseSequentialInvocationContext.java

```diff
@@ -223,118 +257,246 @@ private void invokeNext(VisitableCommand command, Object returnValue, Throwable
          } else {
             // No more interceptors and no more return handlers. We are done!
             if (EXTRA_LOGS && trace)
-               log.tracef("Command %s done, returning %s/%s", command, className(returnValue),
-                     throwable);
+               log.tracef("Command %s done with return value/exception %s/%s", command,
+                     className(returnValue), throwable);
             completeFuture(future, returnValue, throwable);
             return;
          }
       }
    }
 
+   @SuppressWarnings("unchecked")
    private CompletableFuture<Object> handleForkReturn(ForkInfo forkInfo, Object returnValue,
          Throwable throwable) throws Throwable {
-      if (EXTRA_LOGS && trace) {
-         log.tracef("Forked command %s done for interceptor %s, returning %s/%s", forkInfo.newCommand,
-               className(forkInfo.savedInterceptor.interceptor), className(returnValue),
-               throwable);
-      }
       nextInterceptor = forkInfo.savedInterceptor;
       // We are abusing type erasure so that we can handle the future in invokeNext
       CompletableFuture handlerFuture = forkInfo.forkReturnHandler
             .handle(this, forkInfo.savedCommand, returnValue, throwable);
       return handlerFuture;
    }
 
-   Object invokeSync(VisitableCommand command, InterceptorListNode firstInterceptor) throws Throwable {
-      future = null;
-      nextInterceptor = firstInterceptor;
+   Object invokeSync(VisitableCommand command, InterceptorListNode firstInterceptor)
+         throws Throwable {
       action = INVOKE_NEXT;
-
-      return doInvokeNextSync(command, firstInterceptor);
+      try {
+         Object returnValue = invokeInterceptorsSync(command, firstInterceptor);
+         return invokeReturnHandlersSync(command, returnValue, null);
+      } catch (Throwable t) {
+         // Unwrap the exception from CompletableFutures.await
+         Throwable throwable = extractCompletableFutureException(t);
+         return invokeReturnHandlersSync(command, null, throwable);
+      }
    }
 
-   private Object doInvokeNextSync(VisitableCommand command, InterceptorListNode interceptorNode)
+   private Object invokeInterceptorsSync(VisitableCommand command, InterceptorListNode firstInterceptorNode)
          throws Throwable {
+      InterceptorListNode interceptorNode = firstInterceptorNode;
       SequentialInterceptor interceptor = interceptorNode.interceptor;
-      nextInterceptor = interceptorNode.nextNode;
+      interceptorNode = beforeVisit(command, interceptorNode, interceptor);
```

Comment:
I had to manually unroll the loop to get acceptable performance in local mode... adding some comments.